### DISCUSSION TITLE
More python examples?

### DISCUSSION THREAD

**Discussion by JaPhi** [sn4k3/UVtools#550] [Scripts]
Hey everyone,

for a research idea I would like to split some layers into multiple similar to this idea https://blog.honzamrazek.cz/2022/06/preventing-warping-of-resin-printed-pieces-alternative-way-of-fighting-resin-shrinkage. Now I am not super good with C# and was delighted to see the option of using python. Right now there is only the "print_layers.py"-script. I'd love to see some more elaborate examples. Especially transforming the layers into a (x,y,z) numpy array, performing some work on it and putting it back into the proper file format.

I would like to try some different patterns to reduce warping. Maybe with a lot more sub-layer images. Like "white noise".

Oh, right now I am not even getting the print_layers.py to work. "No module named 'pythonnet'" even if it is installed with pip and/or conda. Weird...

**Comment by sn4k3** [comment]
**I highly recommend C#** because it's UVtools native language, can run inside it, it's faster, easy language and you can copy code from it own. After some tries you will get used to it. I saw non programmer people able to do thier own little script.
As bonus you will have code inspection and completion with a proper IDE (this you wont get with python since it will not know UVtools namespace to give you hints) and can be painfull to code without it (You must have source code always open to check function names and arguments).

**Python-wise** i can add more samples (just tell me something you would like as sample), but keep in mind it will always return objects and you will have to call it functions exactly like you would do with C#, i'm not a Python programer so i don't know what you can do with numpy. I added Python for experienced programers but those need to consult the source code and call the functions, the most annoying part as you won't have any code completion and analiser won't brother either with possible errors or typos.

> Oh, right now I am not even getting the print_layers.py to work. "No module named 'pythonnet'" even if it is installed with pip and/or conda. Weird...

You must install from latest dev version, follow this: https://github.com/sn4k3/UVtools/tree/master/Scripts/UVtools.Python it should work if you produce that steps

**Comment by JaPhi** [comment]
The plan is to create some complex pixel movements, and I would greatly benefit from being able to program them in Python. If that is not possible, maybe you are right and I should start looking at C#. I would probably just create an interface to python though. I don't want to relearn how to solve a diffusion equation in c# :D 

if i could make a wish, most of it would revolve around my plan, of course. However, to be further considered as general examples it would be helpful to:

- extract all layers to python (ideally to numpy arrays)
- show how to modify/add/insert layers freely
- change the per-layer settings like exposure time/ wait time etc.
- save such modified layers to .ctb or any other file format
- pause the print after a specific layer - continue by button (if the printer has this feature)

btw. do you know if there is a limit to how fast you can change the image? Would it be possible to basically play animations on each layer?

**Reply by sn4k3** [reply]
> extract all layers to python (ideally to numpy arrays)

The **print_layers.py** sample already loop all layers, you can use that and put them on a numpy array, but you wont benefit from it, in fact it will be slower to apply lambda operations over the Mat than using a for loop. Also you need parallel operations, if you use sequential your operations will take ages to complete if file contain large layer set. In addition you don't want to have all layer images in memory you simply don't have RAM for that so you are stuck with processing in batch.

> show how to modify/add/insert layers freely
change the per-layer settings like exposure time/ wait time etc.
save such modified layers to .ctb or any other file format

All of that is already on [C# samples](https://github.com/sn4k3/UVtools/tree/master/UVtools.ScriptSample), read `ScriptExecute()` function and then all calls are exactly the same on Python, because a C# object on Python must call the same way.
For example take [ScriptChangeLayerPropertiesSample.cs](https://github.com/sn4k3/UVtools/blob/master/UVtools.ScriptSample/ScriptChangeLayerPropertiesSample.cs) as sample

```C#
public bool ScriptExecute()
    {
        Progress.Reset("Changing layers", Operation.LayerRangeCount); // Sets the progress name and number of items to process

        Random random = new();

        for (uint layerIndex = Operation.LayerIndexStart; layerIndex <= Operation.LayerIndexEnd; layerIndex++)
        {
            Progress.ThrowIfCancellationRequested(); // Abort operation, user requested cancellation
            var layer = SlicerFile[layerIndex]; // Unpack and expose layer variable for easier use

            layer.LiftHeight = random.Next(3, 10);     // Random value from 3 to 10
            layer.LiftSpeed = random.Next(50, 200);    // Random value from 50 to 200
            layer.RetractSpeed = random.Next(50, 200); // Random value from 50 to 200

            Progress++; // Increment progress bar by 1
        }

        // return true if not cancelled by user
        return !Progress.Token.IsCancellationRequested;
    }
```

To translate to Python and using the print_layers.py would be:

```python
for layer in slicerFile:
    layer.LiftHeight = random.randint(3, 10)     # Random value from 3 to 10
    layer.LiftSpeed = random.randint(50, 200)    # Random value from 50 to 200
    layer.RetractSpeed = random.randint(50, 200) # Random value from 50 to 200
```

Then you read [Layer.cs](https://github.com/sn4k3/UVtools/blob/master/UVtools.Core/Layers/Layer.cs) to see what you can call or get/set

That way you can take advantage of C# samples and maybe use them as exercices, take C# samples and convert them to Python, that way you can learn UVtools name and functions better. Still you need to look at library sourcecode to call what you really need.

Save a file is easy as `slicerFile.Save` or `slicerFile.SaveAs`, still you want to look at [docs](https://github.com/sn4k3/UVtools/blob/master/UVtools.Core/FileFormats/FileFormat.cs#L4578) to see the arguments

> pause the print after a specific layer - continue by button (if the printer has this feature)

Not possible, printers must support this via the file, none so far support that.

> btw. do you know if there is a limit to how fast you can change the image? Would it be possible to basically play animations on each layer?

On the printer? You will need a gcode printer to get rid of obligatory lifts, but printers still use cheap boards that have a high delay to show image, my gcode printer render a image in 2s to 3s. So you can't make a video with that framerate.
